### PR TITLE
feat: add ws client adapter 

### LINKED ISSUE #259: Add support for WebSocket clients
#### Reason/Context

Currently, Glee only supports creating a WebSocket server. That makes it unusable when someone wants to build a WebSocket client to connect to an existing WS server.

#### Description

We already have support for WebSocket but it's always assuming we want to create a server, and that's plain wrong. Instead, Glee should default to creating a WS client and only create a server when we mark the `server` definition in the AsyncAPI file as "local".

For instance:

```yaml
asyncapi: 2.3.0
servers:
  myWebsiteWebSocketServer:
    url: 'ws://mywebsite.com/ws'
    protocol: ws
    x-kind: local
  someProductWSServer:
    url: 'ws://someproduct.com/ws'
    protocol: ws
```

The server `myWebsiteWebSocketServer` is a WS server we want to create with Glee but the server `someProductWSServer` is a remote server we want to send/receive information to/from. The former is what's already implemented and the latter is what we should be implementing as part of this issue. Notice the `x-kind: local` on the `myWebsiteWebSocketServer` server. Only when this is found, we should create a server, otherwise, we should default to a client. Another allowed value for `x-kind` is `remote`.

#### Notes

`x-kind` is an extension we're creating as part of this issue, this is not defined anywhere yet so we have some freedom defining how we want it to be. I'm trying to make it as similar as possible to the [Remotes and local servers issue on the spec](https://github.com/asyncapi/spec/issues/689). In v3, we may end up having two root keys: `servers` (local servers) and `remotes` (remote servers or brokers). For v2.x (the current version), I think it's just fine if we add the `x-kind` extension but I'm happy to listen for alternatives.

### PR DESCRIPTION
**Pull request by Souvikns**
**Description**
This PR is trying to implement a `websocket` client adapter for glee. 

So the main idea is, if I have a WebSocket server, my glee application should be able to - 
- connect to that server 
- receive and send data
- validate the data received or sent according to the specification file. 

**Related issue(s)**

Fixes #259

### CODE REVIEW

File: src/adapters/ws/client.ts
Comment (Souvikns) [inline src/adapters/ws/client.ts:54]: @fmvilas I am kinda stuck here. 

Currently, we create adapters according to the number of servers we define in the specification file. This works well in the case of servers as we can parse the URL from the request body to figure out which channel to forward the message but in the case of the client, multiple channels could subscribe to different WebSocket servers or to different paths in the same server (something I am doing in my example). 

Can I somehow reference server in channel, I think this is something v3 will be bringing, but right now how do we determine the channel.

File: src/adapters/ws/client.ts
Comment (fmvilas) [inline src/adapters/ws/client.ts:24]: You may want to make this private as well. When I did the first adapters, JS didn't support private stuff on classes 😄

File: src/adapters/ws/client.ts
Comment (fmvilas) [inline src/adapters/ws/client.ts:27]: This is not being used so it means you're not using the channels information at all. Check my comment below 👇

File: src/adapters/ws/client.ts
Comment (fmvilas) [reply src/adapters/ws/client.ts:54]: Replied in other lines to point you in the right direction.

File: src/registerAdapters.ts
Comment (fmvilas) [inline src/registerAdapters.ts:47]: I think we should default to client whenever possible. Glee is a client for multiple protocols. It should be an exception that it behaves as a server IMHO.

**Review by fmvilas — CHANGES_REQUESTED** [review]

File: src/adapters/ws/client.ts
Comment (Souvikns) [reply src/adapters/ws/client.ts:38]: I thought of doing this, but since glee could be a server or a client at the same time -

we could have spec like this 

```yaml
asyncapi: 3.0.0
servers:
  client:
    url: ws: localhost:3001
    protocol: ws
  local: 
    url: ws:localhost:4000
    protocol: ws
    x-kind: local
channels:
  user/signedup: 
    ...
  testClient:
    ...
```

Now how do we figure out which server these channels belong to? or is it bad practice to have more than one server per `asyncapi` file, because I think if my glee application is able to connect to other ws servers then I can connect to multiple servers at once.

Comment (fmvilas) [reply src/adapters/ws/client.ts]: First of all, this example is easy, you should ignore the server that has `x-kind: local` because you don't want to create a client for a server that's "local". If it's local, then we should create a server, not a client.

That said, the AsyncAPI file above, means that `user/signedup` and `testClient` exist in all severs. If you want to restrict channels to certain servers, you can do it using the [`servers` attribute in the Channel Item Object](https://www.asyncapi.com/docs/reference/specification/v2.4.0#channelItemObject).

**Comment by Souvikns** [comment]
@fmvilas need your review again

**Comment by jonaslagoni** [comment]
@fmvilas as you started the review process, I am assuming you wish to finish it 🙂

File: .eslintignore
Comment (fmvilas) [inline .eslintignore:3]: If you end up needing the `ws-server` example, I recommend you move this line to a `.eslintignore` file inside `examples/ws-server`. Probably we should do the same with the line above.

Comment (Souvikns) [reply .eslintignore]: I am removing the `ws-server` example as I don't need it anymore.

**Comment by Souvikns** [comment]
One weird issue I am running into while sending headers from the client side, is the headers are converted to lowercase. 

```ts
{
  'sec-websocket-version': '13',
  'sec-websocket-key': 'nL+pNzTS+0ryHcEU/4ccPQ==',
  connection: 'Upgrade',
  upgrade: 'websocket',
  authorization: 'bearer arb-tokenValue',
  'sec-websocket-extensions': 'permessage-deflate; client_max_window_bits',
  host: 'localhost:3000'
}
```
Whereas I am sending the `Authorization` header and it is received as `authorization`
```yaml
    bindings:
      ws:
        bindingVersion: 0.1.0
        headerValues:
          Authorization: 'bearer $TOKEN'
```

Found this https://stackoverflow.com/questions/50477233/issue-with-netty-websocket-headers-name-in-lowercase

**Review by fmvilas — CHANGES_REQUESTED** [review]

File: examples/crypto-websockets/client/asyncapi.yaml
Comment (fmvilas) [inline examples/crypto-websockets/client/asyncapi.yaml:11]: Let's move this property out of the server definition because that's how it will be done in v3.0.0. For now, we can just use our own extension like this:

```suggestion
x-remoteServers:
  - websockets
```

This will change the logic in the `registerAdapters.ts` file.

File: examples/crypto-websockets/server/asyncapi.yaml
Comment (fmvilas) [reply examples/crypto-websockets/server/asyncapi.yaml:11]: Let's move this property out of the server definition because that's how it will be done in v3.0.0. For now, we can just use our own extension like this:

```suggestion
x-localServers:
  - websocket
```

This will change the logic in the `registerAdapters.ts` file.

**Review by fmvilas — CHANGES_REQUESTED** [review]
Almost there. Just left a suggestion to make remote/local server stuff more similar to what it would be in v3.

File: src/registerAdapters.ts
Comment (fmvilas) [inline src/registerAdapters.ts:29]:
```suggestion
  const remoteServers = parsedAsyncAPI.extension('x-remoteServers')
```

File: src/adapters/ws/client.ts
Comment (fmvilas) [inline src/adapters/ws/client.ts:41]: Not used:

```suggestion
```

**Review by fmvilas — CHANGES_REQUESTED** [review]

File: src/lib/index.d.ts
Comment (Souvikns) [inline src/lib/index.d.ts:22]:
```suggestion
   headers?: any
```

Instead of adding authentication should we just allow header and query config, with headers the user is able to set any type of authentication he desires?

Comment (fmvilas) [reply src/lib/index.d.ts:22]: But the user can already do that once you merge [this PR of yours](https://github.com/asyncapi/bindings/pull/144/files), right? We only saw the need for authentication because it usually contains secrets.

Comment (Souvikns) [reply src/lib/index.d.ts:22]: Yeah, just a suggestion, currently we have an `authentication` config and we are creating a header `Authentication: bearer TOKEN`, and it works but maybe we just allow headers instead of authentication, and then we have a broader control, but we only have the need for authentication. We can always change it in the future if there is a need.

Comment (fmvilas) [reply src/lib/index.d.ts:22]: We'd have broader control at Glee but we'd be breaking the purpose of the AsyncAPI specification. If the `headers` values are in the AsyncAPI document, other tools will be able to see them. If they're in the Glee config, only Glee will be able to see them. You see the problem?

For instance, in #261 we want to use Generator to generate docs automatically. If the info is not in the AsyncAPI file, it will not end up in the documentation. And the same will happen for all the tools that the user might want to apply to their AsyncAPI file.

**Review by fmvilas — APPROVED** [review]
LGTM 👍 🚀

**Comment by Souvikns** [comment]
@fmvilas sentiment analysis is failing, can you rerun the tests, I can't merge right now.

**Comment by fmvilas** [comment]
@Souvikns done. I hope your sentiments are fine now 😜

**Pull request merged**
